ocaml/libs/tracing/tracing.ml

```diff
@@ -0,0 +1,674 @@
+        let content_of (spans : Span.t list) =
+          List.map zipkin_span_of_span spans |> ZipkinSpan.json_of_t_list
+      end
+    end
+  end
+
+  module Destination = struct
+    module File = struct
+      let trace_log_dir = ref "/var/log/dt/zipkinv2/json"
+
+      let host_id = ref "localhost"
+
+      let set_trace_log_dir dir = trace_log_dir := dir
+
+      let set_host_id id = host_id := id
+
+      let export ~trace_id ~span_json ~path : (string, exn) result =
+        try
+          let date = Ptime_clock.now () |> Ptime.to_rfc3339 ~frac_s:6 in
+          let file =
+            String.concat "/" [path; trace_id; "xapi"; !host_id; date] ^ ".json"
+          in
+          Xapi_stdext_unix.Unixext.mkdir_rec (Filename.dirname file) 0o700 ;
+          Xapi_stdext_unix.Unixext.write_string_to_file file span_json ;
+          Ok ""
+        with e -> Error e
+    end
+
+    module Http = struct
+      module Request = Cohttp.Request.Make (Cohttp_posix_io.Buffered_IO)
+      module Response = Cohttp.Response.Make (Cohttp_posix_io.Buffered_IO)
+
+      let export ~span_json ~url : (string, exn) result =
+        try
+          let body = span_json in
+          let uri = Uri.of_string url in
+          let headers =
+            Cohttp.Header.of_list
+              [
+                ("accepts", "application/json")
+              ; ("content-type", "application/json")
+              ; ("content-length", string_of_int (String.length body))
+              ]
+          in
+          Open_uri.with_open_uri uri (fun fd ->
+              let request =
+                Cohttp.Request.make ~meth:`POST ~version:`HTTP_1_1 ~headers uri
+              in
+              let ic = Unix.in_channel_of_descr fd in
+              let oc = Unix.out_channel_of_descr fd in
+              Request.write
+                (fun writer -> Request.write_body writer body)
+                request oc ;
+              Unix.shutdown fd Unix.SHUTDOWN_SEND ;
+              match try Response.read ic with _ -> `Eof with
+              | `Eof ->
+                  Ok ""
+              | `Invalid x ->
+                  Error (Failure ("invalid read: " ^ x))
+              | `Ok response ->
+                  let body = Buffer.create 128 in
+                  let reader = Response.make_body_reader response ic in
+                  let rec loop () =
+                    match Response.read_body_chunk reader with
+                    | Cohttp.Transfer.Chunk x ->
+                        Buffer.add_string body x ; loop ()
+                    | Cohttp.Transfer.Final_chunk x ->
+                        Buffer.add_string body x
+                    | Cohttp.Transfer.Done ->
+                        ()
+                  in
+                  loop () ;
+                  Ok (Buffer.contents body)
+          )
+        with e -> Error e
+    end
+
+    let export_to_endpoint endpoint =
+      try
+        debug "Tracing: About to export" ;
+        let span_list = Spans.since () in
```

Comment:
This gets all newly finished spans, since the last export, right? It doesn't seem to specify that the spans are recorded under a certain tracer or tracer provider, where the endpoint is defined. So if you have multiple tracer providers with different endpoints, then all the same traces are sent to the first endpoint in the `List.iter` below, and the following ones get nothing?

Reply:
Resolved by[ ](https://github.com/xapi-project/xen-api/pull/4958/commits/be140243d2fad8b9f8f4863742a0de5e4700dbfb)

Reply:
https://github.com/xapi-project/xen-api/pull/4958/commits/be140243d2fad8b9f8f4863742a0de5e4700dbfb